lib/cocoapods-core/cdn_source.rb

```diff
@@ -0,0 +1,210 @@
+        Pod::UI.message "CDN: #{name} Relative path: #{partial_url} modified during this run! Returning local"
+        return partial_url
+      end
+
+      path.dirname.mkpath
+
+      etag_path = path.sub_ext(path.extname + '.etag')
+
+      etag = File.read(etag_path) if File.exist?(etag_path)
+      Pod::UI.message "CDN: #{name} Relative path: #{partial_url}, has ETag? #{etag}" unless etag.nil?
+
+      response = REST.get(file_remote_url,
+                          'If-None-Match' => etag)
+      case response.status_code
+      when 304
+        Pod::UI.message "CDN: #{name} Relative path not modified: #{partial_url}"
+        # We need to update the file modification date, as it is later used for freshness
+        # optimization. See #initialize for more information.
+        FileUtils.touch path
+        partial_url
+      when 200
+        File.open(path, 'w') { |f| f.write(response.body) }
+
+        etag_new = response.headers['etag'].first if response.headers.include?('etag')
+        Pod::UI.message "CDN: #{name} Relative path downloaded: #{partial_url}, save ETag: #{etag_new}"
+        File.open(etag_path, 'w') { |f| f.write(etag_new) } unless etag_new.nil?
+        partial_url
+      end
+    end
+
+    def local_file(partial_url)
+      File.open("#{repo}/#{partial_url}") do |file|
+        yield file if block_given?
+      end
+    end
+
+    def local_file_exist?(partial_url)
+      File.exist?("#{repo}/#{partial_url}")
+    end
+
+    def pod_path_partial(pod_name)
+      pod_path(pod_name).relative_path_from(repo)
+    end
+
+    # @!group Querying the source
+    #-------------------------------------------------------------------------#
+
+    # @return [Array<String>] the list of the name of all the Pods.
+    #
+    #
+    def pods
+      download_file('all_pods.txt')
+      local_file('all_pods.txt', &:to_a)
+    end
+
+    # @return [Array<Version>] all the available versions for the Pod, sorted
+    #         from highest to lowest.
+    #
+    # @param  [String] name
+    #         the name of the Pod.
+    #
+    def versions(name)
+      return nil unless specs_dir
+      raise ArgumentError, 'No name' unless name
+
+      return @versions_by_name[name] unless @versions_by_name[name].nil?
+
+      versions_file_name = pod_path_partial(name).join('index.txt')
+      download_file(versions_file_name)
+
+      loaders = []
+      @versions_by_name[name] ||= local_file(versions_file_name) do |file|
+        file.map do |v|
+          basename = v.chomp
+          # Optimization: ensure all the podspec files at least exist. The correct one will get refreshed
+          # in #specification_path regardless.
+          unless local_file_exist?(pod_path_partial(name).join("#{basename}/#{name}.podspec.json"))
+            loaders << Concurrent::Promise.execute(:executor => @executor) do
+              download_file(pod_path_partial(name).join("#{basename}/#{name}.podspec.json"))
```

Comment:
`"#{basename}/#{name}.podspec.json"` could be pulled out into a variable for reuse

Reply:
`pod_path_partial(name)` is also duplicated a few times